Enterprise_Dbcache_WpdbInjection_Cluster.php

```diff
@@ -349,56 +390,61 @@ function db_connect( $query = '', $use_master = null ) {
 
 		if ( $use_master ) {
 			$this->dbhname = $dataset . '__w';
-			$operation = 'write';
+			$operation     = 'write';
 		} else {
 			$this->dbhname = $dataset . '__r';
-			$operation = 'read';
+			$operation     = 'read';
 		}
 
-		// Try to reuse an existing connection
+		// Try to reuse an existing connection.
 		$dbh = $this->_db_connect_reuse_connection();
 		if ( is_object( $dbh ) ) {
 			$this->wpdb_mixin->dbh = $dbh;
 			return $dbh;
 		}
 
-		if ( empty( $this->_cluster_servers[$dataset][$operation] ) )
+		if ( empty( $this->_cluster_servers[ $dataset ][ $operation ] ) ) {
 			return $this->wpdb_mixin->bail( "No databases available for dataset $dataset operation $operation" );
+		}
 
 		// Make a list of at least $this->min_tries connections to try, repeating as necessary.
-		$servers = array();
+		$servers      = array();
+		$server_count = 0;
 		do {
 			foreach ( $this->_current_zone['zone_priorities'] as $zone ) {
-				if ( isset( $this->_cluster_servers[$dataset][$operation][$zone] ) ) {
-					$zone_servers = $this->_cluster_servers[$dataset][$operation][$zone];
+				if ( isset( $this->_cluster_servers[ $dataset ][ $operation ][ $zone ] ) ) {
+					$zone_servers = $this->_cluster_servers[ $dataset ][ $operation ][ $zone ];
 
 					if ( is_array( $zone_servers ) ) {
 						$indexes = array_keys( $zone_servers );
 						shuffle( $indexes );
-						foreach ( $indexes as $index )
+						foreach ( $indexes as $index ) {
 							$servers[] = compact( 'zone', 'index' );
+						}
 					}
 				}
 			}
-		} while ( count( $servers ) < $this->min_tries );
-		// Connect to a database server
+			$server_count = count( $servers );
+		} while ( $server_count < $this->min_tries );
```

Comment:
This will cause an indefinite loop.  The `while` condition must check the `count` of `$servers` for every iteration.  You can change it to `while ( true )` and `break` if the condition is met.

Reply:
After speaking with you, you indicated that this should be okay but could still result in an infinite loop. To resolve this I added a counter and condition to the loop that will reattempt 3 times before ending if the number of servers is inadequate